utilities/backup/backup_engine.cc

```diff
@@ -1432,7 +1433,7 @@ IOStatus BackupEngineImpl::CreateNewBackupWithMetadata(
               live_dst_paths, backup_items_to_finish, new_backup_id,
               false /* shared */, "" /* src_dir */, fname,
               EnvOptions() /* src_env_options */, rate_limiter, type,
-              contents.size(), db_options.statistics.get(), 0 /* size_limit */,
+              contents.size(), db_options.statistics.get(), contents.size(),
```

Comment:
I doubt whether this change is needed for this PR for the following reasons:
(1) `backup_engine_test.cc` passes with reverting this change 
(2) Based on code search, I believe `create_file_cb` is only used for [creating the CURRENT file ](https://github.com/facebook/rocksdb/blob/7.2.fb/utilities/checkpoint/checkpoint_impl.cc#L237-L247) by writing the `info.replacement_contents` with no `SequentialFileReader:Read()` involved. So it goes down [the "Create" path ](https://github.com/facebook/rocksdb/blob/7.2.fb/utilities/backup/backup_engine.cc#L2059)in "CopyAndCreateFile".
(3) Based on the semantics, `create_file_cb` does not seem to call `ReadFileAndComputeChecksum` as it does not involving reading a file hence no checksum verification is needed.  So the change here does not seem to affect the `ReadFileAndComputeChecksum`
(4) Instead, the [one](https://github.com/facebook/rocksdb/blob/d110e17ffb2c1f25990108de814717f99a329a8b/utilities/backup/backup_engine.cc#L1978) under VerifyBackup might be more concerning in terms of over-charging, although the current test does not reveal that. It's probably due to the current test that accidentally revealed overcharging problem was written (by me) for cases involving both read and write rate-limiting while `VerifyBackup` is pure read read-limiting.

But tracing through the test `TEST_P(BackupEngineRateLimitingTestWithParam, RateLimitingVerifyBackup)` + sum up the charged [bytes buffer_to_read](https://github.com/facebook/rocksdb/blob/d110e17ffb2c1f25990108de814717f99a329a8b/utilities/backup/backup_engine.cc#L2374) should reveal that. 

If we really want to make sure your improvement to prevent overcharging is helping, then maybe we should add some unit tests to compare "the requested bytes (i.e, RateLimiter::GetTotalBytesThrough())" VS "the estimated actual read bytes" (how to get this info in test is a bit tricky - maybe Env::GetFileSize()?). I expect to see a relatively big gap between this two before your improvement and a smaller gap (if not exactly equal) after your improvement.

For (4), let me double check with @cbi42 : is `TEST_P(BackupEngineRateLimitingTestWithParam, RateLimitingChargeReadInRestore)` the only test failed without your improvement, which I believe can be solved by changing the test to compare `GetTotalRequests` instead of `GetTotalBytesThrough`? If that's the case, I wouldn't mind saving your over-charging improvement for a separate PR with unit-test and getting the tests passed for this PR (considering it's already big). 

@ajkr  - do you have an opinion on my suggestion above?

Reply:
Btw, @pdillinger can confirm on my understanding about BackupEngine in (2)/(3).
 
Some quick context for you Peter: @cbi42 is working on making BackupEngine cap SequentialFileReader::Read() by file size so that the rate-limiting made inside SequentialFileReader by this PR does not charge more than file size (i.e, the over-charging problem). 

And SequentialFileReader::Read() is called in BackupEngineImpl::ReadFileAndComputeChecksum and BackupEngineImpl::CopyOrCreateFile

Reply:
@hx235 Yes BackupEngineRateLimitingTestWithParam.RateLimitingChargeReadInRestore is the only test that's failing before [this line](https://github.com/facebook/rocksdb/blob/d110e17ffb2c1f25990108de814717f99a329a8b/utilities/backup/backup_engine.cc#L1861) was updated with `file_info->size`.

You are right that `contents.size()` here is not needed, at least for passing all the current test cases. I've reverted the change here and will follow up with a separate PR with some new unit tests.

Reply:
> @ajkr - do you have any opinion on my suggestions above?

 sounds good to me. I am glad if we move the rate limiting to the lower layer.

The code looks good to me. Will defer to Hui to accept since she knows the most detail.